aio.core/aio/core/directory/abstract/directory.py

```diff
@@ -0,0 +1,203 @@
+        return str(self.path.joinpath(path))
+
+    def absolute_paths(self, paths: Iterable[str]) -> Set[str]:
+        """Make directory-relative paths absolute."""
+        return set(
+            self.absolute_path(path)
+            for path
+            in paths)
+
+    def grep(
+            self,
+            args: Iterable,
+            target: Optional[str] = None) -> AwaitableGenerator:
+        return AwaitableGenerator(
+            self._grep(args, target),
+            collector=async_set,
+            predicate=self._include_file)
+
+    def parse_grep_args(
+            self,
+            args: Iterable[str],
+            target: Optional[str] = None) -> Tuple[str, ...]:
+        """Parse `grep` args with the defaults to pass to the `grep`
+        command."""
+        return (
+            *self.grep_args,
+            *args,
+            *((target, )
+              if target is not None
+              else ()))
+
+    @lru_cache
+    def relative_path(self, path: str) -> str:
+        """Make an absolute path directory-relative."""
+        return str(pathlib.Path(path).relative_to(self.path))
+
+    def relative_paths(self, paths: Iterable[str]) -> Set[str]:
+        """Make absolute paths directory-relative."""
+        return set(
+            self.relative_path(path)
+            for path
+            in paths)
+
+    async def _grep(
+            self,
+            args: Iterable,
+            target: Optional[str] = None) -> AsyncIterator[str]:
+        """Run `grep` in the directory."""
+        for path in await self.shell(self.parse_grep_args(args, target)):
+            if path:
+                yield path
+
+    @lru_cache
+    def _include_file(self, path: str) -> bool:
+        return bool(
+            (not self.path_matcher
+             or self.path_matcher.match(path))
+            and (not self.exclude_matcher
+                 or not self.exclude_matcher.match(path)))
+
+
+class AGitDirectory(ADirectory):
+    """A filesystem directory, with an associated fileset, and tools for
+    finding files.
+
+    Uses the `git` index cache for faster grepping.
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        self.changed = kwargs.pop("changed", None)
+        super().__init__(*args, **kwargs)
+
+    @async_property
+    async def changed_files(self) -> Set[str]:
+        """Files that have changed since `self.changed`, which can be the name
+        of a git object (branch etc) or a commit hash."""
+        return set(
+            path
+            for path
+            in await self.shell(
+                [self.git_command,
+                 "diff",
+                 "--name-only",
+                 self.changed,
+                 "--diff-filter=ACMR",
+                 "--ignore-submodules=all"])
+            if (path and self._include_file(path)))
+
+    @async_property(cache=True)
+    async def files(self) -> Set[str]:
+        files = await super().files
```

Comment:
if neither path_matcher nor exclude_matcher are set then its quicker to get the changed_files first